src/blog/memory-safe-languages/myths-zero-overhead-memory-safety.vmd

```diff
@@ -0,0 +1,308 @@
+////
+<div><center><img src="/images/lovebird.jpg" style="max-height: 200px;"/><div style="opacity: .8">(the more _mythical_ birds are below)</div></center></div>
+>>>>
+
+
+In Arrrlang, everything lived in a global array.
+
+ * Every `Ship` was stored in a global `ships` array.
+ * Every `Engine` was stored in a global `engines` array.
+
+...and so on.
+
+
+<<<<
+In Arrrlang, there were no pointers!
+
+
+If a `Ship` wanted to point to its `Engine`, it would instead index into the global `engines` array.
+////
+```
+struct Ship {
+  engineIndex int;
+}
+```
+>>>>
+
+
+Everything was indexing into everything else. It actually felt like using a relational database.
+
+
+With these basic rules, *Arrrlang was memory safe.*
+
+
+# Arrrlang's Memory Safety Costs
+
+Arrrlang had no reference counting, no tracing garbage collection, and no generations. So by Mike's definition, Arrrlang had zero-overhead memory safety.
+
+
+But of course, it had plenty of costs.
+
+ * It's memory usage kept growing over time.
+ * It did a bounds check on every "dereference".
+ * It stored released elements in a free-list, which has some costly cache misses.
+
+
+It might be tempting to think that these are inevitable costs, fundamental to any form of managing memory. 
+
+However, tracing garbage collection doesn't have these costs. [#gccost] It has other costs, but the point still stands.
+
+
+So it seems this zero-overhead memory-safe language actually still has some run-time overhead in practice.
+
+
+If we're looking for a language with zero-overhead memory safety, our expedition apparently isn't over.
+
+
+<div><center><img src="/images/roc.jpg" style="max-height: 200px;"/><div style="opacity: .8">It seems zero-overhead memory safety is as elusive as the<br/>ancient Roc, which was known to carry off entire elephants!</div></center></div>
+
+
+<slice>
+#gccost: Specifically:
+
+ * GC does "compaction" where it moves objects in memory, so that it can release more memory to the OS.
+ * GC's references were just references, and references need no bounds checking.
+ * GCs allocate from a bump allocator, which is very cache-friendly.
+</slice>
+
+# Rust, Bounds Checking, Hashing, Cloning
+
+Rust is in the top three favorites for both me and Mike, so it came up pretty quickly in our conversation.
+
+
+The first thought is, of course, that Rust has bounds checking too, which are an obvious run-time cost (anywhere [between 1-15%](https://shnatsel.medium.com/how-to-avoid-bounds-checks-in-rust-without-unsafe-f65e618b4c1e), usually on the lower side of that).
+
+But all languages have bounds checking! Let's [steelman](https://en.wiktionary.org/wiki/steelman) the question and instead ask, "Does Rust have *more* bounds checking costs *than other languages?*"
```

Comment:
In a sense more vs less bounds checks is a false dichotomy.
Rust lets you skip bounds checks if you want, with a bit of `unsafe`
See https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get_unchecked